### DISCUSSION TITLE
Coding component for J3.x & J4 installation

### DISCUSSION THREAD

**Discussion by skurvish** [joomla/joomla-cms#31906] [Q&A]
With the change to moving namespaced classes to the src directory for components, I am struggling with how this will impact a component we wish to install on both J3 & J4. I reference this PR: https://github.com/joomla/joomla-cms/pull/27687

Are we going to have to have 2 versions of the component, one for each J version? 

Will a folder structure with the src folder install properly on J3.9 or J3.10?

I believe that J4 will allow a J3 component (properly modified) to install but that still means that at some point the component will need to be restructured to match the J4 folder structure.

Am I right to be struggling with this?

**Comment by skurvish** [comment]
Nobody?

**Comment by chmst** [comment]
No, you must not have 2 versions of your component. Your assumption is correct. If you now use the new folder structure with src folder, it will not run on j3. You must  (may) not adapt yout code to the new folder structure, as long as J 3.x is supported.  
This is what I see and have experienced with my components and what I see from the Joomla Code - maybe one of the developers can add more information.

**Comment by skurvish** [comment]
That is what I was afraid of. It means I cannot refactor my component today to run on both J3 & J4 without the requirement sometime down the road that I will have to refactor again. Presumably at some point in time the coding structure that works on both versions will become deprecated and I will be forced to invest in reworking the structure and all of the potential issues a modification of that magnitude might entail. And of course all of the extra effort involved.

**Reply by chmst** [reply]
As there are extensions which come form J1.5 and have always been only adapted to work with newer versions this is not a big risk. Also PHP Versions change and software must be adapted - at the moment to PHP 8. So it is a question of years, and always the same always the same since software exists. For me it is nothing new, I spent my whole professional life with adapting software to newer versions of systems, libraries and programming languages. :)

**Reply by laoneo** [reply]
For sure you can refactor the extension to run on both versions. I do the same with my extensions. All you have to do is to replace the deprecated code from version 3 with the correct one as it is very likely that the deprecated code get removed in J4. If you want to use the new features from version 4, then you need to add some "if" statements in your component. But you are right, you are forced to use the old folder structure from version 3 as only this one runs on both versions.

**Reply by skurvish** [reply]
Thank you both for your clarification. Not what I wanted to hear but it is what it is.
